### PR TITLE
refactor snark0

### PR DESCRIPTION
**Pull request by fabrizio-m**

### CODE REVIEW

File: src/base/snark0.ml [hunk 1]
```diff
@@ -77,387 +80,7 @@ struct
...
-        let ( ||| ) = ( || )
-
-        let not t = Not t
-
-        let any xs = Or xs
-
-        let all xs = And xs
-      end
-    end
+    include Utils.Make (Backend) (Checked) (As_prover_old) (Runner)
```
Comment (rbonichon) [inline src/base/snark0.ml:83]: Looks like you could use the `As prover` module directly here or am I mistaken?

File: src/base/utils.ml [hunk 2]
```diff
@@ -1,48 +1,474 @@
 open Core_kernel
-
-(** Alias [Cvar] to avoid shadowing. *)
+module Types0 = Types
 module Cvar0 = Cvar
+module Checked_ast = Checked_ast
```
Comment (rbonichon) [inline src/base/utils.ml:4]: I'm not sure this re-definition is useful. It even looks like `Checked_ast` is not used in the rest of the file.

File: src/base/snark0.ml [hunk 3]
```diff
@@ -52,6 +52,9 @@ struct
     let field_vars = Array.map fields ~f:(fun x -> Cvar0.Constant x) in
     typ.var_of_fields (field_vars, aux)
 
+  (* to pass to Utils.Make *)
+  module As_prover_old = As_prover
+
```
Comment (rbonichon) [inline src/base/snark0.ml:57]: Not sure this alias is useful (see next comment)

File: src/base/utils.ml [hunk 4]
```diff
@@ -1,48 +1,474 @@
 open Core_kernel
-
-(** Alias [Cvar] to avoid shadowing. *)
+module Types0 = Types
```
Comment (rbonichon) [inline src/base/utils.ml:2]: Searching the file for `Types0` returns no occurrence. Is this alias necessary?

File: src/base/utils.ml [hunk 5]
```diff
@@ -1,48 +1,474 @@
...
-    let%map z = As_prover.read_var z in
-    if Field.equal z Field.zero then (Field.one, Field.zero)
-    else (Field.zero, Field.inv z)
+  module Boolean = Private.Boolean
+
+  let equal, mul, square, div, inv = Private.(equal, mul, square, div, inv)
+
+  let if_, assert_non_zero = Private.(if_, assert_non_zero)
+
+  let equal_vars, equal_constraints = Private.(equal_vars, equal_constraints)
```
Comment (rbonichon) [inline src/base/utils.ml:473]: I'm not sure what the aim here is. 
Every function of this module is public for there is no module signature constraining what's exposed by the `Make` functor. 

If we want function aliases defined at toplevel, we could basically do `include Private` and we'd have the same aliases (and way more). 

If we want to limit what's available within the API, we'd like to do it either though defining a module type, or by defining the module interface (the sibling `.mli` file, which does not currently exist)

**Review by rbonichon — CHANGES_REQUESTED** [review]

File: src/base/snark0.ml (hunk 1 above)
Comment (fabrizio-m) [reply src/base/snark0.ml:83]: Indeed, I was trying to solve another thing and it's no longer necessary.

File: src/base/utils.ml (hunk 2 above)
Comment (fabrizio-m) [reply src/base/utils.ml:4]: It probably is not, I copied most of the definitions from `snark0` when I was not sure how much I was going to port, I'll delete the unneeded ones.

File: src/base/snark0.ml (hunk 3 above)
Comment (fabrizio-m) [reply src/base/snark0.ml:57]: Will remove.

File: src/base/utils.ml (hunk 4 above)
Comment (fabrizio-m) [reply src/base/utils.ml:2]: Will remove.

File: src/base/utils.ml (hunk 5 above)
Comment (fabrizio-m) [reply src/base/utils.ml:473]: I just wanted to limit while working, I will make an .mli

File: src/base/utils.ml [hunk 6]
```diff
@@ -45,4 +69,392 @@ struct
...
+  (* end *)
+
+  (* module Boolean = Private.Boolean *)
+  module Typ2 = Typ
+
+  (* let equal, mul, square, div, inv = Private.(equal, mul, square, div, inv) *)
+
+  (* let if_, assert_non_zero = Private.(if_, assert_non_zero) *)
+
+  (* let equal_vars, equal_constraints = Private.(equal_vars, equal_constraints) *)
```
Comment (rbonichon) [inline src/base/utils.ml:459]: Remove

File: src/base/utils.ml [hunk 7]
```diff
@@ -45,4 +69,392 @@ struct
...
+      let not t = Not t
+
+      let any xs = Or xs
+
+      let all xs = And xs
+    end
+  end
+  (* end *)
+
+  (* module Boolean = Private.Boolean *)
```
Comment (rbonichon) [inline src/base/utils.ml:452]: Remove

File: src/base/utils.ml [hunk 8]
```diff
@@ -1,21 +1,45 @@
...
 struct
-  open struct
-    module Cvar = Backend.Cvar
-    module Field = Backend.Field
-    module Constraint = Backend.Constraint
+  open Backend
+
+  open Runners.Make (Backend) (Checked) (As_prover) (Runner)
+
+  (* module Private = struct *)
```
Comment (rbonichon) [inline src/base/utils.ml:24]: Remove

File: src/base/utils.ml [hunk 9]
```diff
@@ -45,4 +69,392 @@ struct
...
+      let ( ||| ) = ( || )
+
+      let not t = Not t
+
+      let any xs = Or xs
+
+      let all xs = And xs
+    end
+  end
+  (* end *)
```
Comment (rbonichon) [inline src/base/utils.ml:450]: Remove the line

**Review by rbonichon — APPROVED** [review]

**Pull request merged**